runtime-modules/content/src/vnft_auction/types.rs

```diff
@@ -0,0 +1,279 @@
+use super::*;
+
+/// Metadata for vNFT issuance
+type Metadata = Vec<u8>;
+
+pub type CuratorGroupId<T> = <T as ContentActorAuthenticator>::CuratorGroupId;
+pub type CuratorId<T> = <T as ContentActorAuthenticator>::CuratorId;
+pub type MemberId<T> = <T as MembershipTypes>::MemberId;
+
+/// Owner royalty
+pub type Royalty = Perbill;
+
+/// NFT transactional status
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum TransactionalStatus<
+    AccountId,
+    Moment: BaseArithmetic + Copy,
+    CuratorGroupId: Default + Copy,
+    CuratorId: Default + Copy,
+    MemberId: Default + Copy,
+    Balance,
+> {
+    Idle,
+    PendingTransferTo(MemberId),
+    Auction(AuctionRecord<AccountId, Moment, CuratorGroupId, CuratorId, MemberId, Balance>),
+}
+
+impl<
+        AccountId,
+        Moment: BaseArithmetic + Copy,
+        CuratorGroupId: Default + Copy,
+        CuratorId: Default + Copy,
+        MemberId: Default + Copy,
+        Balance,
+    > Default
+    for TransactionalStatus<AccountId, Moment, CuratorGroupId, CuratorId, MemberId, Balance>
+{
+    fn default() -> Self {
+        Self::Idle
+    }
+}
+
+impl<
+        AccountId: Default + PartialEq,
+        Moment: BaseArithmetic + Copy,
+        CuratorGroupId: Default + Copy,
+        CuratorId: Default + Copy,
+        MemberId: Default + Copy,
+        Balance,
+    > OwnedNFT<AccountId, Moment, CuratorGroupId, CuratorId, MemberId, Balance>
+{
+    /// Create new vNFT
+    pub fn new(owner: AccountId, creator_royalty: Option<(AccountId, Royalty)>) -> Self {
+        Self {
+            owner,
+            transactional_status: TransactionalStatus::Idle,
+            creator_royalty,
+        }
+    }
+
+    /// Ensure given account id is vNFT owner
+    pub fn ensure_ownership<T: Trait>(&self, owner: &AccountId) -> DispatchResult {
+        ensure!(self.owner.eq(owner), Error::<T>::DoesNotOwnVNFT);
+        Ok(())
+    }
+}
+
+/// Owned vNFT representation
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct OwnedNFT<
+    AccountId,
+    Moment: BaseArithmetic + Copy,
+    CuratorGroupId: Default + Copy,
+    CuratorId: Default + Copy,
+    MemberId: Default + Copy,
+    Balance,
+> {
+    pub owner: AccountId,
```

Comment:
Good point, but how will we pay to a Member automatically by the end of nft auction? Should we introduce an additional runtime storage to hold total redeemable value for specific members?

Reply:
This happens by the winner triggering `complete_video_auction`, does it not?

I actually looked at that extrinsic in more detail now, and it does not appear to check that the caller is actually the winner, also the call does not statically avoid a successful extrinsic even in cases where there is no NFT auction or if the auction is not expired. These checks should be done, and return `Error`, if not satisfied, before the `mutation safe` barrier.

Reply:
That a good point, if we restrict `complete_video_auction` extrinsic to be called by winner only, which looks logically consistent and correct, we do not need `AccountId`.

Reply:
Thinking about this now. 
Even though we can replace an auction winner `AccountId` with `MemberId`, we still have to hold an `AccountId` of previous owner to be able to complete a payment when auction completes.
What kind of issue do you see here?

Reply:
Also i wouldn't be able to unreserve/unlock previously reserved bid amount.
```
  // Unreserve previous bidder balance
                    T::Currency::unreserve(&last_bid.bidder, last_bid.amount);
```

Reply:
> Even though we can replace an auction winner AccountId with MemberId, we still have to hold an AccountId of previous owner to be able to complete a payment when auction completes.

The payment would go to the controller account of the membership, thats typically the account we use both for paying and receiving payments.

> Also i wouldn't be able to unreserve/unlock bid amount of previous bidder.

When calling `complete_video_auction` which succeeds, the previous bidder is the same as the caller, i.e. the winner.

Reply:
> When calling complete_video_auction which succeeds, the previous bidder is the same as the caller, i.e. the winner.
>
You'll be unable to `unreserve`  amounts of all previous bidders.

Reply:
which we do during `make_bid` extrinisic execution.

Reply:
> You'll be unable to unreserve amounts of all previous bidders.

We only have one bid active at any given time, since they cannot withdraw, a new bid replaces the existing bid if its greater, and thus makes the unreservation. This is how these types of auctions work, the last bid is binding, no older bids are relevant. If the last bid expires in a perpetual auction, then by def. any older bid would have expired also. So there is only one bid life at most.

Even without this property, I actually don't see what exact problem there would be with unreserving all old bids, except for the cost of the possibly very large transaction which can fail, but I don't think its this you are referring to.

One problem that can occur however, which I did not think of, is what if someone changes their membership controller account while having an active bid, in this case, those funds risk getting locked perpetually, and also the unreservation may fail on the new account set. This probably means we need to lock in the actual controller account when making bids, however the owner should still be a member, not a raw account.

Reply:
> We only have one bid active at any given time, since they cannot withdraw, a new bid replaces the existing bid if its greater, and thus makes the unreservation. This is how these types of auctions work, the last bid is binding, no older bids are relevant. If the last bid expires in a perpetual auction, then by def. any older bid would have expired also. So there is > only one bid life at most.
 > 
It's understandable. I previously thought that it's impossible to retrieve controller_account_id for any given `member_id`.
Now, i see, that it's possible.

Reply:
> One problem that can occur however, which I did not think of, is what if someone changes their membership controller account while having an active bid, in this case, those funds risk getting locked perpetually, and also the unreservation may fail on the new account set. This probably means we need to lock in the actual controller account when making bids, > however the owner should still be a member, not a raw account.
Ok, makes sense